daemon/logger/syslog/syslog.go

```diff
@@ -0,0 +1,102 @@
+package syslog
+
+import (
+	"crypto/tls"
+
+	"github.com/alibaba/pouch/daemon/logger"
+	"github.com/alibaba/pouch/daemon/logger/loggerutils"
+
+	"github.com/RackSec/srslog"
+)
+
+// Syslog writes the log data into syslog.
+type Syslog struct {
+	writer *srslog.Writer
+}
+
+type options struct {
+	tag       string
+	proto     string
+	address   string
+	priority  srslog.Priority
+	formatter srslog.Formatter
+	framer    srslog.Framer
+	tlsCfg    *tls.Config
+}
+
+func defaultOptions() *options {
+	return &options{
+		priority: defaultSyslogPriority,
+	}
+}
+
+// NewSyslog returns new Syslog based on the log config.
+func NewSyslog(info logger.Info) (*Syslog, error) {
+	opts, err := parseOptions(info)
+	if err != nil {
+		return nil, err
+	}
+
+	var w *srslog.Writer
+	if opts.proto == secureProto {
+		w, err = srslog.DialWithTLSConfig(opts.proto, opts.address, opts.priority, opts.tag, opts.tlsCfg)
+	} else {
+		w, err = srslog.Dial(opts.proto, opts.address, opts.priority, opts.tag)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	w.SetFormatter(opts.formatter)
+	w.SetFramer(opts.framer)
+	return &Syslog{writer: w}, nil
+}
+
+// WriteLogMessage will write the LogMessage.
+func (s *Syslog) WriteLogMessage(msg *logger.LogMessage) error {
+	line := string(msg.Line)
+	if msg.Source == "stderr" {
+		return s.writer.Err(line)
+	}
+	return s.writer.Info(line)
+}
+
+// Close closes the Syslog.
+func (s *Syslog) Close() error {
+	return s.writer.Close()
```

Comment:
Is it OK to close Syslog multiple times?

Reply:
In my opinion, I don't think we should care the multiple closes. Even if the srslog supports the case. F

We should close the syslog once. If we closes it multiple times, we should reconsider that the logic is ok. And that we will introduce sync.Lock things if we support the case. Is it worth? 😄

Reply:
Yes, I totally agree that the backend should be closed only once.

But now pouchd will include all the backends in both `Stdout` and `Stderr`. when being closed, both of them will close all the backends. And this is unreasonable.

So my PR #1388 fixes this problem. The close of `Stdout` and `Stderr` only close the corresponding ring buffer. Only when the `IO` is closed, all the backends will be closed 😄

Reply:
@YaoZengzeng we need to drain the ringbuffer, right? When we close the IO, we need to wait the data has been drained. 😄

Reply:
Yes, it is exactly what #1388 do. 😄